spec/statsd/udp_connection_spec.rb

```diff
@@ -580,3 +580,35 @@
     end
   end
 end
+
+
+# This instance is only about testing when the connection is opened
+
+describe Datadog::Statsd::UDPConnection do
+  subject do
+    described_class.new(host, port, logger: logger, telemetry: telemetry)
+  end
+  let(:host) do
+    'fakedomain'
+  end
+  let(:port) do
+    4567
+  end
+  let(:logger) do
+    Logger.new(STDOUT)
+  end
+  let(:telemetry) do
+    instance_double(Datadog::Statsd::Telemetry, sent: true, dropped: true)
+  end
+  describe '#initialize' do
+    it 'is not immediately opening the connection' do
+      expect(subject).not_to receive(:connect)
+    end
+  end
+  describe '#write' do
+    it 'is opening the connection' do
+      expect(subject).to receive(:connect)
+      subject.write("hello")
+    end
+  end
+end
```

Comment:
I believe this can be simplified and replaced with a 

```ruby
    it 'does not immediately connect' do
      expect(UDPSocket).to_not receive(:new)

      subject
    end
```

under `describe '#initialize'`. If we forget to `connect` before calling `write`, a bunch of other testcases fail anyway, so all we need to do is ensure that it doesn't happen in the constructor.

The same note applies to `uds_connection_spec.rb`.

(Also just as an rspec usage note, nested `describe`/`context` blocks allow you to redefine only the `let`s you need to and inherit the others, so if we wanted to keep these exact tests, we could merge them into the ones above using that technique. Afaik what rspec does is create a class hierarchy where the describe/contexts create a new subclass of the one they are nested inside of -- you can even call `super()` inside a let and it will allow you to reference the parent let, if you want to do weirder things).

Reply:
HA! I've tried to do that first but I was not looking at the `UDPSocket`, I was trying to observe the `:connect` in `UDPConnection` and I did not succeed to do so (was giving false positive on `master`).

Thanks for the hints!!

Reply:
There's this really good presentation by Sandi Metz where she talks about these kinds of tests and in particular to avoid doing assertions on calls to the current object, but instead assert on outbound interactions :)

I always remember this slide:

[![Screen Shot 2021-10-20 at 10 20 28](https://user-images.githubusercontent.com/2785847/138065924-81cd23c5-9899-4268-8d0c-e9076d978b60.png)](https://www.youtube.com/watch?v=URSWYvyc42M)

Reply:
Interesting! Thank you 🙇